agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java

```diff
@@ -0,0 +1,261 @@
+        final String path = url.getPath();
+        if (path.isEmpty()) {
+          urlBuilder.append("/");
+        } else {
+          urlBuilder.append(path);
+        }
+        final String query = url.getQuery();
+        if (query != null) {
+          urlBuilder.append("?").append(query);
+        }
+        final String fragment = url.getFragment();
+        if (fragment != null) {
+          urlBuilder.append("#").append(fragment);
+        }
+
+        span.setAttribute(Tags.HTTP_URL, urlBuilder.toString());
+
+        if (Config.get().isHttpServerTagQueryString()) {
+          span.setAttribute(MoreTags.HTTP_QUERY, url.getQuery());
+          span.setAttribute(MoreTags.HTTP_FRAGMENT, url.getFragment());
+        }
+      }
+    } catch (final Exception e) {
+      log.debug("Error tagging url", e);
+    }
+    // TODO set resource name from URL.
+  }
+
+  public boolean sameTrace(Span oneSpan, Span otherSpan) {
+    return oneSpan.getContext().getTraceId().equals(otherSpan.getContext().getTraceId());
+  }
+
+  /**
+   * This method is used to generate an acceptable span (operation) name based on a given method
+   * reference. Anonymous classes are named based on their parent.
+   */
+  protected String spanNameForMethod(final Method method) {
+    return spanNameForClass(method.getDeclaringClass()) + "." + method.getName();
+  }
+
+  /**
+   * This method is used to generate an acceptable span (operation) name based on a given class
+   * reference. Anonymous classes are named based on their parent.
+   */
+  public String spanNameForClass(final Class clazz) {
+    if (!clazz.isAnonymousClass()) {
+      return clazz.getSimpleName();
+    }
+    String className = clazz.getName();
+    if (clazz.getPackage() != null) {
+      final String pkgName = clazz.getPackage().getName();
+      if (!pkgName.isEmpty()) {
+        className = clazz.getName().replace(pkgName, "").substring(1);
+      }
+    }
+    return className;
+  }
+
+  protected void onError(final Span span, final Throwable throwable) {
+    addThrowable(span, unwrapThrowable(throwable));
+  }
+
+  // TODO semantic attributes
+  public static void addThrowable(final Span span, final Throwable throwable) {
+    span.setAttribute(MoreTags.ERROR_MSG, throwable.getMessage());
+    span.setAttribute(MoreTags.ERROR_TYPE, throwable.getClass().getName());
+
+    final StringWriter errorString = new StringWriter();
+    throwable.printStackTrace(new PrintWriter(errorString));
+    span.setAttribute(MoreTags.ERROR_STACK, errorString.toString());
+  }
+
+  public Span getCurrentSpan() {
+    return tracer.getCurrentSpan();
+  }
+
+  public Scope newScope(Span span) {
+    return tracer.withSpan(span);
+  }
+
+  // TODO should end methods remove SPAN attribute from request as well?
+  public void end(Span span, Scope scope, int responseCode) {
+    if (scope == null) {
+      return;
+    }
+
+    if (span != null) {
+      end(span, responseCode);
+    }
+
+    scope.close();
+  }
+
```

Comment:
it doesn't looks like this is ever called with `null` span or `null` scope

also, i think i like calling `end(span, responseStatus)` directly, and keeping `scope.close()` an explicit separate action

```suggestion
```

Reply:
This again will force instrumentation to call several methods on tracer in row. The whole point of these "semantic/typed tracers" for me is encapsulating all those steps that are repeated again and again in every instrumentation. I want as few repetitive lines of code in instrumentation classes as possible.

Reply:
And it can be called with `null` span if instrumentation decided not to create new span. Servlet instrumentations do that from time to time.

Reply:
I have moved scope management out of tracer for now. But I have a strong feeling that #465 will want to bring it back :)